coderd/wsbuilder/wsbuilder.go

```diff
@@ -0,0 +1,741 @@
+// Package wsbuilder provides the Builder object, which encapsulates the common business logic of inserting a new
+// workspace build into the database.
+package wsbuilder
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/google/uuid"
+	"github.com/lib/pq"
+	"github.com/tabbed/pqtype"
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/coderd/conversion"
+	"github.com/coder/coder/coderd/database"
+	"github.com/coder/coder/coderd/httpapi"
+	"github.com/coder/coder/coderd/provisionerdserver"
+	"github.com/coder/coder/coderd/rbac"
+	"github.com/coder/coder/coderd/tracing"
+	"github.com/coder/coder/codersdk"
+)
+
+// Builder encapsulates the business logic of inserting a new workspace build into the database.
+//
+// Builder follows the so-called "Builder" pattern where options that customize the kind of build you get return
+// a new instance of the Builder with the option applied.
+//
+// Example:
+//
+// b = wsbuilder.New(workspace, transition).VersionID(vID).Initiator(me)
+// build, job, err := b.Build(...)
+type Builder struct {
+	// settings that control the kind of build you get
+	workspace             database.Workspace
+	trans                 database.WorkspaceTransition
+	version               versionTarget
+	state                 stateTarget
+	logLevel              string
+	legacyParameterValues []codersdk.CreateParameterRequest
+	richParameterValues   []codersdk.WorkspaceBuildParameter
+	initiator             uuid.UUID
+	reason                database.BuildReason
+
+	// used during build, makes function arguments less verbose
+	ctx   context.Context
+	store database.Store
+
+	// cache of objects, so we only fetch once
+	template                  *database.Template
+	templateVersion           *database.TemplateVersion
+	templateVersionJob        *database.ProvisionerJob
+	templateVersionParameters *[]database.TemplateVersionParameter
+	lastBuild                 *database.WorkspaceBuild
+	lastBuildErr              *error
+	lastBuildParameters       *[]database.WorkspaceBuildParameter
+	lastParameterValues       *[]database.ParameterValue
+	lastBuildJob              *database.ProvisionerJob
+}
+
+type versionTarget struct {
+	active   bool
+	specific *uuid.UUID
+}
+
+type stateTarget struct {
+	orphan   bool
+	explicit *[]byte
+}
+
+func New(w database.Workspace, t database.WorkspaceTransition) Builder {
+	return Builder{workspace: w, trans: t}
+}
+
+// Methods that customize the build are public, have a struct receiver and return a new Builder.
+
+func (b Builder) VersionID(v uuid.UUID) Builder {
+	// nolint: revive
+	b.version = versionTarget{specific: &v}
+	return b
+}
```

Comment:
Can we expose these on the struct directly and remove these helper functions?

Reply:
Then the `New` func could be discarded too if we wanted, which could be nice.

Reply:
We can.

I'll admit I do like the [Fluent interface](https://en.wikipedia.org/wiki/Fluent_interface) that we'd lose.

```
builder := wsbuilder.New(workspace, database.WorkspaceTransitionStart).
    Reason(database.BuildReasonInitiator).
    Initiator(apiKey.UserID).
    ActiveVersion().
    LegacyParameterValues(createWorkspace.ParameterValues).
    RichParameterValues(createWorkspace.RichParameterValues).
    State(state)
```

Would become

```
builder := wsbuilder.Builder{
    Workspace: workspace,
    Transition: database.WorkspaceTransitionStart,
    Reason: database.BuildReasonInitiator,
    Initiator: apiKey.UserID,
    Version: VersionTarget{Active: true},
    LegacyParameterValues: createWorkspace.ParameterValues,
    RichParameterValues: createWorkspace.RichParameterValues,
    State: StateTarget{Explicit: state}
}
```